File: LeetCode/1249.Minimum_Remove_to_Make_Valid_Parentheses/1249.Minimum_Remove_to_Make_Valid_Parentheses.swift
class Solution {
    func minRemoveToMakeValid(_ s: String) -> String {
        // stack
        // if (, push it into stack
        // if ), if stack has (, pop
        var chars = Array(s)
        var stack: [Int] = []

        for (i, ch) in chars.enumerated() {
            if ch == "(" {
                stack.append(i)
            } else if ch == ")" {
                if !stack.isEmpty {
                    stack.removeLast()
                } else {
                    chars[i] = " "
                }
            }
        }

        for index in stack {
            chars[index] = " "
        }

        return String(chars).filter { $0 != " " }
    }
}

File: LeetCode/1249.Minimum_Remove_to_Make_Valid_Parentheses/1249.Minimum_Remove_to_Make_Valid_Parentheses.py
#!/usr/bin/env python3


class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        # 1. empty str, contains only lowercase
        # 2. AB
        # 3. A

        n = len(s)
        res = []
        stack = []  # store index of '('

        for i, c in enumerate(s):
            if c == "(":
                stack.append(i)
                res.append(c)
            elif c == ")":
                if stack:
                    stack.pop()
                    res.append(c)
                else:
                    res.append("")
            else:
                res.append(c)

        for i in stack:
            res[i] = ""

        return ("").join(res)
